core/include/drivers/pinctrl.h

```diff
@@ -0,0 +1,94 @@
+/* SPDX-License-Identifier: BSD-2-Clause */
+/*
+ * Copyright (c) 2022, Microchip
+ */
+
+#ifndef __DRIVERS_PINCTRL_H
+#define __DRIVERS_PINCTRL_H
+
+#include <sys/queue.h>
+#include <tee_api_types.h>
+
+enum pin_config_mode {
+	PIN_MODE_BIAS_DISABLE,
+	PIN_MODE_BIAS_PULL_UP,
+	PIN_MODE_BIAS_PULL_DOWN,
+};
+
+/*
+ * struct pinctrl - Describe a pinctrl controller
+ *
+ * @node: Node offset of the controller in the device-tree
+ * @apply_state: callback which apply the content of a specific pinctrl conf
+ * @link: Link in list of pinctrl
+ */
+struct pinctrl {
+	int node;
+	TEE_Result (*apply_state)(struct pinctrl *pinctrl, const void *fdt,
+				  int node);
+	SLIST_ENTRY(pinctrl) link;
+};
+
+#ifdef CFG_DRIVERS_PINCTRL
+
+/**
+ * pinctrl_register - Register a pin controller
+ *
+ * @pinctrl: Structure representing the pin controller
+ */
+void pinctrl_register(struct pinctrl *pinctrl);
+
+/**
+ * pinctrl_apply_state - Apply a pinctrl state
+ *
+ * @fdt: Device-tree containing the pinctrl node
+ * @node: Node of the device for which to apply the pinctrl state
+ * @name: Name of the pinctrl state to apply or NULL for the "default" one
+ *
+ * Return a TEE_Result compliant value
+ */
+TEE_Result pinctrl_apply_state(const void *fdt, int node, const char *name);
```

Comment:
+1

Reply:
Regarding the dynamic applying of pinctrl, This is something rarely used, even in Linux driver. Do you have such use case ?

About the `DT_DRIVER_PINCTRL` and clk like probing/callbacks, unless the other type of controller that are using `*-cells` properties, pinctrl do not use such notation but rather a phandle to a subnode of the parent pinctrl node. 
Moreover, rather than the `get_of_device` callback is not applicable since  the pinctrl apply function needs a node to be applied and not some `struct dt_driver_phandle_args` that contaisn a few cells value.  So, this method of looking up is not easily applicable to pinctrl nodes unless modifying your existing DT code to handle that exact specific case and passing a node  in the phandle args rather than cells value. Not sure if this is better though.

Regarding` TEE_Result pinctrl_dt_get(fdt, node, struct pinctrl **pinctrl);`, what is this function supposed to do ?  A node might have multiple pinctrl subnode references so I'm not sure what you expect here and in this PR `struct pinctrl` refers to a pin controller. So maybe you meant something else.

Reply:
> Regarding the dynamic applying of pinctrl, This is something rarely used, even in Linux driver. Do you have such use case ?

I am thinking of suspend time here specifc config can be applied by drivers.

> About the DT_DRIVER_PINCTRL and clk like probing/callbacks ...
> ... So, this method of looking up is not easily applicable to pinctrl

Ok, understood, the device lists of dt_driver will not help much.

> Regarding TEE_Result pinctrl_dt_get(fdt, node, struct pinctrl **pinctrl);, what is this function supposed to do ? ...

If I understood correctly your explainations, I would expect it gets the consumer node as input arg and outputs a handle for a pinmiux config list (e.g. `pinconf_list *`) which can be use at run or PM time to load alternate pinmux configs, as argument to `pinconf_load_by_index()`.

Reply:
> > Regarding the dynamic applying of pinctrl, This is something rarely used, even in Linux driver. Do you have such use case ?
> 
> I am thinking of suspend time here specifc config can be applied by drivers.
> 
> > About the DT_DRIVER_PINCTRL and clk like probing/callbacks ...
> > ... So, this method of looking up is not easily applicable to pinctrl
> 
> Ok, understood, the device lists of dt_driver will not help much.
Actually, after thinking a bit about it, we might potentially modify your DT parsing code with the following modification:
- Return 0 as cells count for DT_DRIVER_PINCTRL rather than parsing a cells property
- Add a "int node_phandle" in the `struct dt_driver_phandle_args` which will contains the phandle of the controller/node in case of pinctrl (ie, the phandle specified before the cells values when refering to them)
- In dt device lookup, when resolving the references, in case we handle a DT_DRIVER_PINCTRL, get the parent instead of looking for the phandle to resolve the provider

If you are ok with that, I think that with these modifications, we can get rid of the custom code I have and be integrated with your DT driver framework. There will be some special handling but that seems rather limited to me. I'll try that and resubmit it. 

> > Regarding TEE_Result pinctrl_dt_get(fdt, node, struct pinctrl **pinctrl);, what is this function supposed to do ? ...
> 
> If I understood correctly your explainations, I would expect it gets the consumer node as input arg and outputs a handle for a pinmiux config list (e.g. `pinconf_list *`) which can be use at run or PM time to load alternate pinmux configs, as argument to `pinconf_load_by_index()`.

Ok, get it, I was confuse by the use of the `struct pinctrl` I think this can also be done at the cost of increase use of memory due to some intermediate structure. I think this is also cleaner and will thus match a bit more the use of other frameworks.

Reply:
I think we're aligned. I see we need some clear vocabulary, i am also confusing pinctrl provider refs, pinctrl config refs and pinctrl consumer config-list refs.

As for the DT bnidings that allow to refer to a pinmux config by_name() or by_index(), we could keep it simple and only support by_index() at runtime, with a helper name2index() converter helper function to not malloc/strdup names  at init.